### PR TITLE
Compatible with spring boot 2.x

### LINKED ISSUE #13602: can't compatible spring boot 2.x
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?

5.0.0

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-JDBC

### Actual behavior

upgrade spring boot to 2.x, org.apache.shardingsphere.spring.boot.SpringBootStarterTest#assertRules can't run
image

### Reason analyze (If you can)

org.apache.shardingsphere.spring.boot.util.PropertyUtil#v2 can't worker well

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
1.  only need to upgrade the spring boot in shardingsphere to 2.0.0.RELEASE 
2. org.apache.shardingsphere.spring.boot.SpringBootStarterTest#assertRules

### PR DESCRIPTION
**Pull request by SirMin**
Fixes #13602.

Changes proposed in this pull request:

1. springboot2.x config  key toDashed

### CODE REVIEW

**Review by TeslaCN — COMMENTED** [review]
LICENSE in shardingsphere-distribution should also be updated if versions of dependencies changed.

**Comment by SirMin** [comment]
> LICENSE in shardingsphere-distribution should also be updated if versions of dependencies changed.

already update

**Review by SirMin — COMMENTED** [review]
update LICENSE

**Comment by TeslaCN** [comment]
@zjcnb Please help check this PR.

**Comment by zjcnb** [comment]
> @zjcnb Please help check this PR.

Thx, I will check it.

**Comment by zjcnb** [comment]
@SirMin Hello, Can you pull and merge master branch?

**Comment by SirMin** [comment]
@zjcnb 
> @SirMin Hello, Can you pull and merge master branch?

done

**Comment by zjcnb** [comment]
> 1. upgrade snakeyaml 1.16 -> 1.29
> 2. Expand the release coverage of Springframework
> 3. Expand the release coverage of Spring boot
> 4. Modifying unit tests
> 
> Fixes #13602.
> 
> Changes proposed in this pull request:
> 
> * upgrade snakeyaml 1.16 -> 1.29
> * Expand the release coverage of Springframework
> * Expand the release coverage of Spring boot
> * Modifying unit tests

@SirMin Hi, thank for your submit this pull request, this idea likes good. But i hava a question.

First, Springboot 2.x not support use `_` , it 's only support `-`, i think this Springboot problem. ShardingSphere maybe needn't to compatible it. I think throws exception is better.

**Comment by SirMin** [comment]
@zjcnb springboot only @ConfigurationProperties not support  '_', but @Value support.
code in pr is from spring.
image

org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm.
the springboot version is 2.3.4.RELEASE

**Comment by zjcnb** [comment]
> @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> 
> org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE

Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.

**Comment by SirMin** [comment]
> Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.

@zjcnb spring boot 2.x not support snakeyaml 1.16.

**Comment by SirMin** [comment]
> > @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> > org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE
> 
> Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.

@zjcnb spring boot 2.x not support snakeyaml 1.16. need to uppgrade

**Comment by zjcnb** [comment]
> > > @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> > > org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE
> > 
> > 
> > Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.
> 
> @zjcnb spring boot 2.x not support snakeyaml 1.16. need to uppgrade

I think add toDashedForm() to resolve your problem, springboot version maybe shouldn't upgradle.

**Comment by SirMin** [comment]
> > > > @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> > > > org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE
> > > 
> > > 
> > > Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.
> > 
> > 
> > @zjcnb spring boot 2.x not support snakeyaml 1.16. need to uppgrade
> 
> I think add toDashedForm() to resolve your problem, springboot version maybe shouldn't upgradle.

Does the unit test not cover the springboot 2.x version?

**Comment by zjcnb** [comment]
> > > > > @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> > > > > org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE
> > > > 
> > > > 
> > > > Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.
> > > 
> > > 
> > > @zjcnb spring boot 2.x not support snakeyaml 1.16. need to uppgrade
> > 
> > 
> > I think add toDashedForm() to resolve your problem, springboot version maybe shouldn't upgradle.
> 
> Does the unit test not cover the springboot 2.x version? if not covert I will rollback it.

Yeah, Please rollback about springboot upgradle. only keep toDashedForm().

**Comment by SirMin** [comment]
> > > > > > @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> > > > > > org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE
> > > > > 
> > > > > 
> > > > > Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.
> > > > 
> > > > 
> > > > @zjcnb spring boot 2.x not support snakeyaml 1.16. need to uppgrade
> > > 
> > > 
> > > I think add toDashedForm() to resolve your problem, springboot version maybe shouldn't upgradle.
> > 
> > 
> > Does the unit test not cover the springboot 2.x version? if not covert I will rollback it.
> 
> Yeah, Please rollback about springboot upgradle. only keep toDashedForm().

done

**Comment by zjcnb** [comment]
Thanks. i will check it again.

File: shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-spring-infra/shardingsphere-jdbc-spring-boot-starter-infra/src/main/java/org/apache/shardingsphere/spring/boot/util/PropertyUtil.java [hunk 1]
```diff
@@ -101,13 +101,45 @@ private static Object v1(final Environment environment, final String prefix, fin
...
         Method resultGetMethod = bindResultObject.getClass().getDeclaredMethod("get");
         return resultGetMethod.invoke(bindResultObject);
     }
+
+    /**
+     * Return the specified Java Bean property name in dashed form.
+     * @param name the source name
+     * @return the dashed from
+     */
+    public static String toDashedForm(final String name) {
```
Comment (zjcnb) [inline shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-spring-infra/shardingsphere-jdbc-spring-boot-starter-infra/src/main/java/org/apache/shardingsphere/spring/boot/util/PropertyUtil.java:120]: Change `public` to `private`

Comment (SirMin) [reply shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-spring-infra/shardingsphere-jdbc-spring-boot-starter-infra/src/main/java/org/apache/shardingsphere/spring/boot/util/PropertyUtil.java]: @zjcnb done

**Comment by linghengqian** [comment]
> > > @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> > > org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE
> > 
> > Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.
> 
> @zjcnb spring boot 2.x not support snakeyaml 1.16. need to uppgrade

Does this have much to do with the ` snakeyaml.version` ? I want to know whether the generic lost it with https://github.com/apache/shardingsphere/pull/15260 .

**Comment by SirMin** [comment]
> > > > @zjcnb springboot only @ConfigurationProperties not support '_', but @value support. code in pr is from spring. image
> > > > org.springframework.boot.context.properties.bind.DataObjectPropertyName#toDashedForm. the springboot version is 2.3.4.RELEASE
> > > 
> > > 
> > > Yes, i see. By the way, i think add toDashedForm() for PropertyUtil.class, it's ok. But why change dependency.
> > 
> > 
> > @zjcnb spring boot 2.x not support snakeyaml 1.16. need to uppgrade
> 
> Does this have much to do with the ` snakeyaml.version` ? I want to know whether the generic lost it with #15260 .

@linghengqian yes,  SET type generic lost in new snakeyaml. can use list instead of it,  or set type

**Comment by zjcnb** [comment]
@SirMin This pull request has some conflicts, Can you resolve it?

**Comment by SirMin** [comment]
> @SirMin This pull request has some conflicts, Can you resolve it?

@zjcnb done!
1. Modify the private to default for unit testing 
2. add JUnit test

**Review by menghaoranss — APPROVED** [review]

**Comment by linghengqian** [comment]
> > 
> > Does this have much to do with the ` snakeyaml.version` ? I want to know whether the generic lost it with #15260 .
> 
> @linghengqian yes,  SET type generic lost in new snakeyaml. can use list instead of it,  or set type 

I did not reply to you at the first time because I was busy. This is actually a bug introduced since snakeyaml 1.19 . Reference in https://bitbucket.org/snakeyaml/snakeyaml/issues/387/support-for-generic-types-when-serializing .

**Pull request merged**